src/datasets/arrow_dataset.py

```diff
@@ -1997,6 +1997,11 @@ def init_buffer_and_writer():
         if update_data:
             # Create new Dataset from buffer or file
             info = self.info.copy()
+            # Remove task templates if the required features have been removed
+            if info.task_templates:
+                for idx, task in enumerate(info.task_templates):
+                    if not all(k in writer._features.keys() for k in task.features.keys()):
+                        _ = info.task_templates.pop(idx)
```

Comment:
Using  `pop` here seems dangerous - what if we remove two templates? The index of the second one is not `idx`

Reply:
Personnally, I would have used a list comprehension here:

```python
info.task_templates = [template if all(k in writer._features.keys() for k in template.features.keys()) for template in info.task_templates]
```

Reply:
oh yes you are indeed correct - well spotted! i've now switch to using a list comprehension as you suggested